mcstatus/motd/transformers.py

```diff
@@ -0,0 +1,303 @@
+from __future__ import annotations
+
+import abc
+import typing as t
+from collections.abc import Sequence
+
+from mcstatus.motd.components import Formatting, MinecraftColor, ParsedMotdComponent, TranslationTag, WebColor
+
+
+def _camel_to_snake_case(string: str, /) -> str:
+    """Converts a CamelCase string to snake_case.
+
+    Found at https://stackoverflow.com/a/44969381.
+    """
+    return "".join(["_" + char.lower() if char.isupper() else char for char in string]).lstrip("_")
+
+
+_HOOK_RETURN_TYPE = t.TypeVar("_HOOK_RETURN_TYPE")
+_END_RESULT_TYPE = t.TypeVar("_END_RESULT_TYPE")
+
+
+class BaseTransformer(abc.ABC, t.Generic[_HOOK_RETURN_TYPE, _END_RESULT_TYPE]):
+    """Base motd transformer class.
+
+    Motd transformer is responsible for providing a way to generate an alternative representation
+    of motd, such as one that is able to be printed in the terminal.
+    """
+
+    def transform(self, motd_components: Sequence[ParsedMotdComponent]) -> _END_RESULT_TYPE:
+        return self._format_output([self._handle_component(component) for component in motd_components])
+
+    @abc.abstractmethod
+    def _format_output(self, results: list[_HOOK_RETURN_TYPE]) -> _END_RESULT_TYPE:
+        ...
+
+    def _handle_component(self, component: ParsedMotdComponent) -> _HOOK_RETURN_TYPE:
+        return getattr(self, f"_handle_{_camel_to_snake_case(component.__class__.__name__)}")(component)
```

Comment:
I don't know how I feel about this much dynamicness here. I'd probably prefer to just handle the individual components in an `if-elif-elif` way. This is because this kind of dynamic approach doesn't play well with type-checkers, which basically just have to give the result of `getattr` the `Any` type, avoiding any kind of type checks here.

Reply:
For type checkers, we have abstract methods for every possible type.

Reply:
Still though, I find this too "magical" for my taste. While it does work, it's really not necessary and we shouldn't introduce this much complexity into the code unless we actually need it. Yes, this will mean a bit more repetition in code, but it's not terribly bad and I'd be much happier with that kind of repetition than with this complex logic.

Reply:
Fixed [here](https://github.com/py-mine/mcstatus/compare/2e01e825dc0d40547e1722c481a6ff689628700a..46dabec73e057cf0efac55b8afc100162b03313f), thanks!